### ISSUE TITLE
Modifying context

### ISSUE THREAD

**Issue by cliedeman** [DanielMSchmidt/apollo-opentracing#42]
Due to this [line](https://github.com/DanielMSchmidt/apollo-opentracing/blob/master/src/index.ts#L119). Is it not possible to have race conditions where a resolver may receive the wrong span?

**Comment by cliedeman** [comment]
With the latest changes in `apollo-server` I experimented with attaching a method to the context:

```
class ResolverContext {
  startSpan(info: GraphqlResolveInfo): Span;
}
```

This allowed me to retrieve/create a span based on the (hopefully) unique path in the query. (This uses new extension changes in apollo-server that have not been [released](https://github.com/apollographql/apollo-server/pull/1547)). Do you think that could work?

**Comment by DanielMSchmidt** [comment]
I don't see the benefit of this approach, can you explain it to me? What do you exactly mean by unique path in the query?

**Comment by cliedeman** [comment]
I have created an [example project](https://github.com/cliedeman/apollo-tracing-example) that illustrates the issue:

There are 3 fields that have an artifical delay to ensure they complete out of order.
The [line](https://github.com/cliedeman/apollo-tracing-example/blob/master/src/Extension.ts#L33) simulates modifying the context.

If we execute this query:
```
{
  field1
  field2
  field3
}
```

The following gets logged out:
```
Updating Context field1
Resolving field1, Context: field1
Updating Context field2
Resolving field2, Context: field2
Updating Context field3
Resolving field3, Context: field3
Resolved field2, Context: field3
Resolved field3, Context: field3
Resolved field1, Context: field3
```

So in short in the resolver we *could* access the span from another resolver.
This can be easily circumvented by either saving the span at the start of the resolver or destructuring
`resolve(parent, args, { span}, info)`
but it's easy to forget to do this and mess up the tracing.

I have created #47 to demonstrate an alternative.
